File: src/services/audit/detectors/ad/accounts/patterns/smartcard-not-required.ts
/**
 * Smartcard Not Required Detector
 * Check for accounts without smartcard requirement
 *
 * Detects enabled user accounts that don't have SMARTCARD_REQUIRED flag set.
 * In high-security environments, critical accounts should require smartcard.
 *
 * Note: This is a broad check. For admin-specific detection, use ADMIN_NO_SMARTCARD.
 * UAC flag 0x40000 = SMARTCARD_REQUIRED
 */

import { ADUser } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedUserEntities } from '../../../../../../utils/entity-converter';

export function detectSmartcardNotRequired(users: ADUser[], includeDetails: boolean): Finding {
  // Only check enabled accounts with adminCount=1 (privileged accounts)
  // Regular users without smartcard is expected in most environments
  const affected = users.filter((u) => {
    if (!u.enabled) return false;
    if (!u.adminCount || u.adminCount !== 1) return false;

    const uac = u.userAccountControl || 0;
    // Check if SMARTCARD_REQUIRED is NOT set
    return (uac & 0x40000) === 0;
  });

  return {
    type: 'SMARTCARD_NOT_REQUIRED',
    severity: 'medium',
    category: 'accounts',
    title: 'Smartcard Not Required',
    description:
      'Privileged accounts (adminCount=1) without smartcard requirement. ' +
      'High-value accounts should require strong authentication.',
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedUserEntities(affected) : undefined,
  };
}

File: internal/audit/detectors/ad/accounts/patterns/smartcard-not-required.go
package patterns

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// SmartcardNotRequiredDetector detects accounts without smartcard requirement
type SmartcardNotRequiredDetector struct {
	audit.BaseDetector
}

// NewSmartcardNotRequiredDetector creates a new detector
func NewSmartcardNotRequiredDetector() *SmartcardNotRequiredDetector {
	return &SmartcardNotRequiredDetector{
		BaseDetector: audit.NewBaseDetector("SMARTCARD_NOT_REQUIRED", audit.CategoryAccounts),
	}
}

// Detect executes the detection
func (d *SmartcardNotRequiredDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []types.User

	// UAC flag 0x40000 = SMARTCARD_REQUIRED
	const smartcardRequiredFlag = 0x40000

	for _, u := range data.Users {
		// Only check enabled privileged accounts
		if u.Disabled {
			continue
		}
		if !u.AdminCount {
			continue
		}

		// Check if SMARTCARD_REQUIRED is NOT set
		if (u.UserAccountControl & smartcardRequiredFlag) == 0 {
			affected = append(affected, u)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "Smartcard Not Required",
		Description: "Privileged accounts (adminCount=1) without smartcard requirement. High-value accounts should require strong authentication.",
		Count:       len(affected),
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedUserEntities(affected)
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewSmartcardNotRequiredDetector())
}
